third_party/move/move-vm/runtime/src/storage/loader/traits.rs

```diff
@@ -50,7 +71,66 @@ pub trait ModuleMetadataLoader {
     ) -> PartialVMResult<Vec<Metadata>>;
 }
 
+/// Configuration used by legacy eager loader only. Used to allow single implementation for both
+/// metered and not metered entrypoints like entry functions or scripts.
+pub struct LegacyLoaderConfig {
```

Comment:
Would it help to have some abstraction that combines traversal context and gas meter, as well as some charging configurations. That way instead of passing 2 (traversal context and gas meter) and sometimes 3 (legacy config) everywhere, we might be able to pass a single structure that could also potentially expose some charging APIs (e.g. if charge_for_dependencies / charge_for_ty_tag_dependencies do logic in eager). Maybe think if something along these lines could make sense, or what else could be restructured and if it could be a net win.

Reply:
Yes, gas meter and context probably can be combined together (I would rather o it after landing PRs though ahah). For config: I think the problem is that we have entry points in AptosVM where we do not charge (view functions), where we charge based on gas feature version (entry function), and closures (same code, but we always traverse things)... so it cannot be hidden behind loader, but passed around as you mentioned but it would need repacking / resetting (entrypoint does not charge -> inner function dispatch does charge)

Reply:
Basically we need to have `Wrapper { config, meter, context }` but `config` needs to be reset throughout the execution :/

Reply:
I think it is okay to have an override interface on the wrapper, or maybe config can itself be stateful - that would actually force us to understand and formalize the possible flows (of when to charge, when not to charge and how can it evolve throughout the lifetime of a single config), which can be useful in its own right. Then depending on the finding, we can hide whether to charge or not behind API or just reset the variables and add more state, or reset in the wrapper, or just keep separate :D But yes, we can have a todo/task for this and revisit separately.